### PR TITLE
Add filters to allow themes and plugins to pass HTML attributes to different Nav Walker output

### PR DESCRIPTION
**Pull request by davidwebca**
Currently, Nav Walkers allow custom HTML attributes to be passed to the link (<a>) element only. There are many instances, like when creating a menu with a declarative JavaScript library, where it would be ideal to be able to pass down HTML attributes to any of the built element. UL and LI are currently excluded of this.

Trac ticket: https://core.trac.wordpress.org/ticket/57140

### CODE REVIEW

File: src/wp-includes/class-walker-nav-menu.php [hunk 1]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
-		$class_names = $class_names ? ' class="' . esc_attr( $class_names ) . '"' : '';
 
-		$output .= "{$n}{$indent}<ul$class_names>{$n}";
+		$atts 		   = array();
+		$atts['class'] = ! empty( $class_names ) ? $class_names : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu item's ul element.
+		 *
+		 * @since 6.1.2
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 * @since 6.2.0
```
The earliest that this new filter could be merged to Core is WordPress 6.2.0.

🔢 Applies elsewhere in the PR.

File: src/wp-includes/class-walker-nav-menu.php [hunk 2]
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+	 *
+	 * @param      array   $atts   The atts
+	 * @return     string  HTML attributes string
+	 */
+	protected function build_atts( $atts = array() ) {
+		$attributes = '';
+		foreach ( $atts as $attr => $value ) {
+			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
+				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
+				$attributes .= ' ' . $attr . '="' . $value . '"';
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
				$attributes .= ' ' . esc_attr( $attr ) . '="' . $value . '"';
```
Attribute names will also need to be escaped here.

File: src/wp-includes/class-walker-nav-menu.php [hunk 3]
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+	 *
+	 * @since 6.1.2
+	 *
+	 * @param      array   $atts   The atts
+	 * @return     string  HTML attributes string
+	 */
+	protected function build_atts( $atts = array() ) {
+		$attributes = '';
+		foreach ( $atts as $attr => $value ) {
+			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
			if ( false !== $value && is_scalar( $value ) && '' !== trim( $value ) ) {
```
1. Slight re-ordering to skip early if `$value` is `false`.
2. `trim( $value )` to prevent values with only blank space. - If this isn't preferred here, let me know and include an example of an acceptable empty value to help clarify this for me.

File: src/wp-includes/class-walker-nav-menu.php (hunk 2 above)
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]: Curious about `(bool) true` values here.

`(bool) false` are skipped, but `(bool) true` will be accepted. That is, `$attr = true` will become `attr="1"`.

Is this the intended/desired behaviour, or should `(bool) true` be treated as a boolean attribute and only `esc_attr( $attr )` be appended to `$attributes`?

File: src/wp-includes/class-walker-nav-menu.php [hunk 4]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
 		 * @param int      $depth   Depth of menu item. Used for padding.
 		 */
 		$class_names = implode( ' ', apply_filters( 'nav_menu_submenu_css_class', $classes, $args, $depth ) );
-		$class_names = $class_names ? ' class="' . esc_attr( $class_names ) . '"' : '';
 
-		$output .= "{$n}{$indent}<ul$class_names>{$n}";
+		$atts 		   = array();
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$atts          = array();
```
Should fix this PHPCS failure.

File: src/wp-includes/class-walker-nav-menu.php [hunk 5]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
+		 *
+		 * @param array $atts {
+		 *     The HTML attributes applied to the `<ul>` element, empty strings are ignored.
+		 *
+		 *     @type string $class    HTML CSS class attribute
+		 * }
+		 * @param stdClass $args      An object of `wp_nav_menu()` arguments.
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$atts 		= apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$atts        = apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
```
Should fix this PHPCS failure.

File: src/wp-includes/class-walker-nav-menu.php [hunk 6]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
 		 */
 		$class_names = implode( ' ', apply_filters( 'nav_menu_submenu_css_class', $classes, $args, $depth ) );
-		$class_names = $class_names ? ' class="' . esc_attr( $class_names ) . '"' : '';
 
-		$output .= "{$n}{$indent}<ul$class_names>{$n}";
+		$atts 		   = array();
+		$atts['class'] = ! empty( $class_names ) ? $class_names : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu item's ul element.
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 * Filters the HTML attributes applied to a menu list element.
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 7]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
+		 *
+		 * @since 6.1.2
+		 *
+		 * @param array $atts {
+		 *     The HTML attributes applied to the `<ul>` element, empty strings are ignored.
+		 *
+		 *     @type string $class    HTML CSS class attribute
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 *     @type string $class    HTML CSS class attribute.
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 8]
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
 		$id = apply_filters( 'nav_menu_item_id', 'menu-item-' . $menu_item->ID, $menu_item, $args, $depth );
-		$id = $id ? ' id="' . esc_attr( $id ) . '"' : '';
 
-		$output .= $indent . '<li' . $id . $class_names . '>';
+		$li_atts 		  = array();
+		$li_atts['class'] = ! empty( $class_names ) ? $class_names : '';
+		$li_atts['id']    = ! empty( $id ) ? $id : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu item's li element.
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 * Filters the HTML attributes applied to a menu's list item element.
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 9]
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
+		 *
+		 * @since 6.1.2
+		 *
+		 * @param array $li_atts {
+		 *     The HTML attributes applied to the menu item's `<li>` element, empty strings are ignored.
+		 *
+		 *     @type string $class        HTML CSS class attribute.
+		 *     @type string $id           ID html attribute.
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 *     @type string $id           HTML id attribute.
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 10]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
+		 * @param array $atts {
+		 *     The HTML attributes applied to the `<ul>` element, empty strings are ignored.
+		 *
+		 *     @type string $class    HTML CSS class attribute
+		 * }
+		 * @param stdClass $args      An object of `wp_nav_menu()` arguments.
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$atts 		= apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
+		$attributes = $this->build_atts($atts);
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$attributes = $this->build_atts( $atts );
```
Should fix these PHPCS failures.

File: src/wp-includes/class-walker-nav-menu.php [hunk 11]
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
 		 * @param int      $depth        Depth of menu item. Used for padding.
 		 */
 		$id = apply_filters( 'nav_menu_item_id', 'menu-item-' . $menu_item->ID, $menu_item, $args, $depth );
-		$id = $id ? ' id="' . esc_attr( $id ) . '"' : '';
 
-		$output .= $indent . '<li' . $id . $class_names . '>';
+		$li_atts 		  = array();
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$li_atts          = array();
```
Should fix this PHPCS failure.

File: src/wp-includes/class-walker-nav-menu.php [hunk 12]
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
+		 *
+		 *     @type string $class        HTML CSS class attribute.
+		 *     @type string $id           ID html attribute.
+		 * }
+		 * @param WP_Post  $menu_item The current menu item object.
+		 * @param stdClass $args      An object of wp_nav_menu() arguments.
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$li_atts 	   = apply_filters( 'nav_menu_item_attributes', $li_atts, $menu_item, $args, $depth );
+		$li_attributes = $this->build_atts($li_atts);
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$li_attributes = $this->build_atts( $li_atts );
```
Should fix these PHPCS failures.

File: src/wp-includes/class-walker-nav-menu.php [hunk 13]
```diff
@@ -204,15 +242,8 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
-		$atts = apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
-
-		$attributes = '';
-		foreach ( $atts as $attr => $value ) {
-			if ( is_scalar( $value ) && '' !== $value && false !== $value ) {
-				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
-				$attributes .= ' ' . $attr . '="' . $value . '"';
-			}
-		}
+		$atts 		= apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$atts       = apply_filters( 'nav_menu_link_attributes', $atts, $menu_item, $args, $depth );
```
Should fix this PHPCS failure.

File: src/wp-includes/class-walker-nav-menu.php [hunk 14]
```diff
@@ -204,15 +242,8 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
+		$attributes = $this->build_atts($atts);
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		$attributes = $this->build_atts( $atts );
```
Should fix this PHPCS failures.

File: src/wp-includes/class-walker-nav-menu.php [hunk 15]
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
 		$output .= "</li>{$n}";
 	}
 
+	/**
+	 * Builds an HTML attributes string based on an array of key value pairs
+	 * Empty values are ignored
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
	 * Builds a string of HTML attributes from an array of key/value pairs.
	 * Empty values are ignored.
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 16]
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+	 *
+	 * @since 6.1.2
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
	 * @since 6.2.0
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 17]
```diff
@@ -276,4 +307,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+	 *
+	 * @param      array   $atts   The atts
+	 * @return     string  HTML attributes string
```
Comment (costdev) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
	 * @param array $atts Optional. An array of HTML attribute key/value pairs. Default empty array.
	 * @return string  A string of HTML attributes.
```

**Review by costdev — CHANGES_REQUESTED** [review]
Hi @davidwebca, thanks for the PR!

I've left some suggestions to improve security and documentation, ask a question about boolean attributes, and help resolve the PHPCS failures.

Alternatively, for the PHPCS failures, you could run this in your terminal to autofix them:
```sh
./vendor/bin/phpcbf src/wp-includes/class-walker-nav-menu.php
```

File: src/wp-includes/class-walker-nav-menu.php (hunk 2 above)
Comment (davidwebca) [reply src/wp-includes/class-walker-nav-menu.php]: Hi. I don't think this is applicable here. Attributes values and names have different possible characters so the escaping mechanism should be different. WordPress doesn't have an attribute name escaping mecanism yet so we should avoid adding it here since [it wasn't there before](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215). It could very well introduce breaking change to millions of users. I'll add some links down below to document previous conversations that happened about this in the last 10 years on WP core.

File: src/wp-includes/class-walker-nav-menu.php (hunk 3 above)
Comment (davidwebca) [reply src/wp-includes/class-walker-nav-menu.php]: I agree on the reordering, but blank values are allowed in HTML values so trimming shouldn't be added here. The main thing that I would like to further address in a future patch would be to allow empty html attributes (ex.: "disabled" without any value or <span aria-hidden>Something</span>) This would involve a bit of refactoring. I'll resolve this for now.

File: src/wp-includes/class-walker-nav-menu.php (hunk 2 above)
Comment (davidwebca) [reply src/wp-includes/class-walker-nav-menu.php]: I mainly replicated what was already in use in this file to avoid creating breaking changes as much as possible. Any scalar value can be easily transformed into a string by PHP, some people might already be relying in their code on ```(bool) true``` becoming (string) "1" so I wouldn't want to step into that for now.

**Comment by davidwebca** [comment]
Thanks @costdev for your pointers, very helpful. 

As for the HTML attributes building function, here's my train of thought: 

At first, I started using the [code that was already there](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215) with the assumptions that it had already been approved by core and had the least risk of causing breaking changes for end users. I had replicated the code multiple times and it started feeling silly so I extracted it into its separated protected function in the Nav Walker's class. But, then, my research lead me to previous efforts regarding this: 

- A general HTML attribute building function just like the one I extracted that could be added to formatting.php - https://core.trac.wordpress.org/ticket/23236. Opened 10 years ago and closed 9 years ago for lack of interest.
- A general HTML attribute name escape function with an interesting conversation about the differences in allowed characters to escape or not in attribute names versus values. https://core.trac.wordpress.org/ticket/43010

I think it's safe to assume that the path to have this included as fast as possible without introducing breaking changes in Nav Walkers would be to continue with this patch without involving too many different conversations at once, but once that's done, we could start looking at reviving those two conversations, especially since the second one already has a patch and is just sitting there without activity. It would be amazing to have a general HTML attributes builder functions that every classes could refactor to use and eventually debloat + secure a lot of code.

**Comment by costdev** [comment]
@davidwebca Thanks for the detailed response!

I agree with you that given the following:
1. The existing code doesn't escape attribute names.
2. Extenders may be using spec-valid attribute names that `esc_attr()` would break.
3. Core doesn't have a spec-valid escape for attribute names.

This PR shouldn't escape the attribute names, doesn't introduce an additional security issue here, and escaping these attribute names should be part of wider discussion about the introduction of a spec-valid escape for attribute names.

Had this been newly introduced code, the introduction of a spec-valid escape in future would be an _enhancement_, not a regression. However, using `esc_attr()` in this PR would be a regression.

------

Regarding test failures, note the previous output:
```php
$output .= $indent . '<li' . $id . $class_names . '>';
```

Reversing lines 192 and 193 in the PR shows all tests passing with:
```sh
./vendor/bin/phpunit --filter '/Tests_Menu_Walker_Nav_Menu|Tests_Menu_wpNavMenu/'
```

**Comment by peterwilsoncc** [comment]
> This PR shouldn't escape the attribute names, doesn't introduce an additional security issue here, and escaping these attribute names should be part of wider discussion about the introduction of a spec-valid escape for attribute names.

Just a quick note that WordPress prevents double escaping, so `esc_attr()` should be fine in this context. From `wp shell` to demonstrate:

```
wp> $attr = 'Pens & Pencils';
=> string(14) "Pens & Pencils"
wp> $attr = esc_attr( $attr );
=> string(18) "Pens &amp; Pencils"
wp> $attr = esc_attr( $attr );
=> string(18) "Pens &amp; Pencils"
```

**Comment by costdev** [comment]
@peterwilsoncc So you don't anticipate any issues caused by introducing `esc_attr()` for attribute names in this PR?

If so, should this change be made in this PR, or a follow-up ticket to note the change?

**Comment by peterwilsoncc** [comment]
@costdev No, I don't. I'm happy for the change to remain in this PR, the data is currently escaped and this PR is simply moving around where that takes place. Feel free to correct me if I am misunderstanding, I've only done a quick scan of the PR.

**Comment by costdev** [comment]
Summary of discussion with @peterwilsoncc in Slack:
- [The existing code](https://github.com/WordPress/wordpress-develop/blob/d2f420493d87465c323190892d4e7a6e4c6070a7/src/wp-includes/class-walker-nav-menu.php#L209-L215) doesn't escape attribute names, only values.
- Introducing `esc_attr()` to attribute names here could break CSS styling.
- Further consideration is needed before escaping attribute names in already-existing code.

In short: The PR can continue without introducing `esc_attr()` for attribute names.

**Comment by davidwebca** [comment]
I've made the swap suggested by @costdev and tests seem to be okay.

**Comment by davidwebca** [comment]
Hi @costdev! Just wanted to check on this. Is there anything else I can do to help get the ball rolling or is this just waiting to be pulled in?

**Comment by davidwebca** [comment]
Tests are added. I didn't read them through but since the patch was given by @costdev I have full confidence it's ok. Sorry for the multiple commits, there were some conflicts when applying the patch.

File: tests/phpunit/tests/menu/walker-nav-menu.php [hunk 18]
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
...
+
+		$this->assertSame( 1, $filter->get_call_count() );
+	}
+
+	/**
+	 * Tests that `Walker_Nav_Menu::start_el()` applies 'nav_menu_item_attributes' filters.
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::start_el()
```
Comment (costdev) [inline tests/phpunit/tests/menu/walker-nav-menu.php]:
```suggestion
	 * @covers Walker_Nav_Menu::start_el
```

File: tests/phpunit/tests/menu/walker-nav-menu.php [hunk 19]
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
...
+
+		$this->assertSame( 1, $filter->get_call_count() );
+	}
+
+	/**
+	 * Tests that `Walker_Nav_Menu::build_atts()` builds attributes correctly.
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::build_atts()
```
Comment (costdev) [inline tests/phpunit/tests/menu/walker-nav-menu.php]:
```suggestion
	 * @covers Walker_Nav_Menu::build_atts
```

File: tests/phpunit/tests/menu/walker-nav-menu.php [hunk 20]
```diff
@@ -365,4 +365,117 @@ public function test_walker_nav_menu_start_el_should_add_rel_privacy_policy_when
 
 		$this->assertStringContainsString( 'rel="privacy-policy"', $output );
 	}
+
+	/**
+	 * Tests that `Walker_Nav_Menu::start_lvl()` applies 'nav_menu_submenu_attributes' filters.
+	 *
+	 * @ticket 57278
+	 *
+	 * @covers Walker_Nav_Menu::start_lvl()
```
Comment (costdev) [inline tests/phpunit/tests/menu/walker-nav-menu.php]:
```suggestion
	 * @covers Walker_Nav_Menu::start_lvl
```

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for adding the tests @davidwebca!

I accidentally included `()` in the `@covers` annotations. I've left the suggestions to remove these in this review 🙂

File: src/wp-includes/class-walker-nav-menu.php [hunk 21]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
-		$class_names = $class_names ? ' class="' . esc_attr( $class_names ) . '"' : '';
 
-		$output .= "{$n}{$indent}<ul$class_names>{$n}";
+		$atts          = array();
+		$atts['class'] = ! empty( $class_names ) ? $class_names : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu list element.
+		 *
+		 * @since 6.2.0
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 * @since 6.3.0
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 22]
```diff
@@ -170,9 +187,30 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
...
 
-		$output .= $indent . '<li' . $id . $class_names . '>';
+		$li_atts          = array();
+		$li_atts['id']    = ! empty( $id ) ? $id : '';
+		$li_atts['class'] = ! empty( $class_names ) ? $class_names : '';
+
+		/**
+		 * Filters the HTML attributes applied to a menu's list item element.
+		 *
+		 * @since 6.2.0
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		 * @since 6.3.0
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 23]
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
 		$output .= "</li>{$n}";
 	}
 
+	/**
+	 * Builds a string of HTML attributes from an array of key/value pairs.
+	 * Empty values are ignored.
+	 *
+	 * @since 6.2.0
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
	 * @since 6.3.0
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 24]
```diff
@@ -73,9 +73,27 @@ public function start_lvl( &$output, $depth = 0, $args = null ) {
...
+		 *
+		 *     @type string $class    HTML CSS class attribute.
+		 * }
+		 * @param stdClass $args      An object of `wp_nav_menu()` arguments.
+		 * @param int      $depth     Depth of menu item. Used for padding.
+		 */
+		$atts       = apply_filters( 'nav_menu_submenu_attributes', $atts, $args, $depth );
+		$attributes = $this->build_atts( $atts );
+
+		$output .= "{$n}{$indent}<ul$attributes>{$n}";
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]: Nit: for consistency within the string.

```suggestion
		$output .= "{$n}{$indent}<ul{$attributes}>{$n}";
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 25]
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+	 * Builds a string of HTML attributes from an array of key/value pairs.
+	 * Empty values are ignored.
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param  array $atts Optional. An array of HTML attribute key/value pairs. Default empty array.
+	 * @return string A string of HTML attributes.
+	 */
+	protected function build_atts( $atts = array() ) {
+		$attributes = '';
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]: Having `$atts`, `$attributes` and `$attr` all within one function is a little unclear, let's change the outcome to something clearer. (I know this was an existing issue but it's good to improve while moving the code.)

```suggestion
		$attribute_string = '';
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 26]
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+		foreach ( $atts as $attr => $value ) {
+			if ( false !== $value && '' !== $value && is_scalar( $value ) ) {
+				$value       = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
+				$attributes .= ' ' . $attr . '="' . $value . '"';
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
				$value             = ( 'href' === $attr ) ? esc_url( $value ) : esc_attr( $value );
				$attribute_string .= ' ' . $attr . '="' . $value . '"';
```

File: src/wp-includes/class-walker-nav-menu.php [hunk 27]
```diff
@@ -286,4 +317,24 @@ public function end_el( &$output, $data_object, $depth = 0, $args = null ) {
...
+			}
+		}
+		return $attributes;
```
Comment (peterwilsoncc) [inline src/wp-includes/class-walker-nav-menu.php]:
```suggestion
		return $attribute_string;
```

**Review by peterwilsoncc — COMMENTED** [review]
I've added a few notes inline but they're all minor.

File: src/wp-includes/class-walker-nav-menu.php (hunk 25 above)
Comment (costdev) [reply src/wp-includes/class-walker-nav-menu.php:330]: By the same token I'd consider renaming the method to `build_attributes()` in the context of "you don't build bricks/$atts, you build houses/$attributes". Besides, `build_attributes` requires no expansion, so it's cleaner.

**Comment by davidwebca** [comment]
Thanks for all the suggestions. I've added them now.

**Comment by peterwilsoncc** [comment]
Committed in [r56067](https://core.trac.wordpress.org/changeset/56067)
